### PR TITLE
Convert Spring Web MVC library instrumentation to Instrumenter API

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Part of https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2713

### CODE REVIEW

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcNetAttributesExtractor.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    extends NetAttributesExtractor<HttpServletRequest, HttpServletResponse> {
+  @Override
+  public String transport(HttpServletRequest request) {
+    return SemanticAttributes.NetTransportValues.IP_TCP;
+  }
+
+  @Override
+  public @Nullable String peerName(
+      HttpServletRequest request, @Nullable HttpServletResponse response) {
+    return null;
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcNetAttributesExtractor.java:24]: Could be filled with
```suggestion
    return request.getRemoteHost();
```

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+  @Override
+  protected @Nullable String flavor(
+      HttpServletRequest request, @Nullable HttpServletResponse response) {
+    return request.getProtocol();
+  }
+
+  @Override
+  protected Integer statusCode(HttpServletRequest request, HttpServletResponse response) {
+    return response.getStatus();
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java:45]: Would it make sense to mimic servlet instrumentation and set status to 500 when response isn't committed and there is an exception?

File: instrumentation/spring/spring-webmvc-3.1/library/README.md [hunk 3]
```diff
@@ -79,12 +79,12 @@ import org.springframework.web.client.RestTemplate;
 public class WebMvcTracingFilterConfig {
 
    @Bean
-   public WebMvcTracingFilter webMvcTracingFilter(Tracer tracer) {
-      return new WebMvcTracingFilter(tracer);
+   public WebMvcTracingFilter webMvcTracingFilter(OpenTelemetry openTelemetry) {
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc-3.1/library/README.md:82]:
```suggestion
   public Filter webMvcTracingFilter(OpenTelemetry openTelemetry) {
```

Comment (laurit) [reply instrumentation/spring/spring-webmvc-3.1/library/README.md]: I believe that out of the box this construct works only for spring boot, for plain spring mvc you'll need to also add the filter to web.xml. There is a sample in instrumentation/spring/spring-webmvc-3.1/wildfly-testing/src/test/resources/web.xml

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java:45]: Can you point me to the class where it happens?

File: instrumentation/spring/spring-webmvc-3.1/library/README.md (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-3.1/library/README.md]: Yeah, that may be true - the whole instrumentations seems to be written with spring boot (and our spring-boot-autoconfigure module) in mind.
I'll add it to the README file.

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  public SpringWebMvcTracing build() {
+    SpringWebMvcHttpAttributesExtractor httpAttributesExtractor =
+        new SpringWebMvcHttpAttributesExtractor();
+
+    Instrumenter<HttpServletRequest, HttpServletResponse> instrumenter =
+        Instrumenter.<HttpServletRequest, HttpServletResponse>newBuilder(
+                openTelemetry,
+                INSTRUMENTATION_NAME,
+                HttpSpanNameExtractor.create(httpAttributesExtractor))
```
Comment (laurit) [inline instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java:55]: If I understand correctly this just sets span name to `HTTP GET`, is it correct to say that this fixes #3479? I guess we could somehow adapt `OpenTelemetryHandlerMappingFilter` for library instrumentation that should give better names.

**Review by laurit — APPROVED** [review]

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java (hunk 2 above)
Comment (laurit) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java:45]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d55eee2ebf74b21ba9f63da44d511d7316d43de8/instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/ServletHttpAttributesExtractor.java#L112
To pass throwable response is wrapped, which happens here
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d55eee2ebf74b21ba9f63da44d511d7316d43de8/instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/ServletHelper.java#L51

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java:55]: > If I understand correctly this just sets span name to `HTTP GET`, is it correct to say that this fixes #3479?

🤷‍♂️  It's better than the class name I guess, and it fits the HTTP spec.
We can't really get better names on the `Filter` level, we'd probably have to implement some additional piece of code that injects itself into Spring machinery - I'm not really sure whether this is worth the effort though.

Comment (laurit) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java:55]: Although it is out of the scope of this pr we already have this beast https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java A filter that deduces which spring mapping is going to match and sets the name of the server span. I guess the intention behind it is that if request is denied by spring security then it doesn't reach to actual matching so to get a good span name we need to figure out what matches the request before the actual matching in spring takes place. The same approach could probably also be used in library instrumentation.

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java (hunk 2 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcHttpAttributesExtractor.java:45]: Nice catch, I'll add this. Thanks!

File: instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/SpringWebMvcTracingBuilder.java:55]: That's wild, but I think it's doable. I'll remove the association with #3479 and link your comment there - it's a good idea to at least try it out. Thanks!

**Review by trask — APPROVED** [review]

**Pull request merged**